susemanager-utils/susemanager-sls/modules/roster/uyuni.py

```diff
@@ -0,0 +1,309 @@
+        SELECT ENCODE(SHA256(FORMAT('%s|%s|%s|%s|%s|%s',
+                      EXTRACT(EPOCH FROM MAX(S.modified)),
+                      COUNT(S.id),
+                      EXTRACT(EPOCH FROM MAX(SP.modified)),
+                      COUNT(SP.proxy_server_id),
+                      EXTRACT(EPOCH FROM MAX(SMI.modified)),
+                      COUNT(SMI.server_id)
+               )::bytea), 'hex') AS fp
+               FROM rhnServer AS S
+               INNER JOIN suseMinionInfo AS SMI ON
+                     (SMI.server_id=S.id)
+               LEFT JOIN rhnServerPath AS SP ON
+                    (SP.server_id=S.id)
+               WHERE S.contact_method_id IN (
+                         SELECT SSCM.id
+                         FROM suseServerContactMethod AS SSCM
+                         WHERE SSCM.label IN ('ssh-push', 'ssh-push-tunnel')
+                     )
+    """
+    h = _executeQuery(query)
+    if h is not None:
+        row = h.fetchone()
+        if row and row[0]:
+            new_fp = row[0]
+            if new_fp == cache_fp and "minions" in cache_data and cache_data["minions"]:
+                log.debug("Return the cached data")
+                return __utils__["roster_matcher.targets"](
+                    cache_data["minions"], tgt, tgt_type
+                )
+            else:
+                log.debug("Invalidate cache")
+                cache_fp = new_fp
+                ret = {}
+    else:
+        log.warning(
+            "Unable to reconnect to the Uyuni DB. Returning cached data instead."
+        )
+        return __utils__["roster_matcher.targets"](cache_data["minions"], tgt, tgt_type)
+
+    query = """
+        SELECT S.id AS server_id,
+               SMI.minion_id AS minion_id,
+               SMI.ssh_push_port AS ssh_push_port,
+               SSCM.label='ssh-push-tunnel' AS tunnel,
+               SP.hostname AS proxy_hostname
+        FROM rhnServer AS S
+        INNER JOIN suseServerContactMethod AS SSCM ON
+              (SSCM.id=S.contact_method_id)
+        INNER JOIN suseMinionInfo AS SMI ON
+              (SMI.server_id=S.id)
+        LEFT JOIN rhnServerPath AS SP ON
+             (SP.server_id=S.id)
+        WHERE SSCM.label IN ('ssh-push', 'ssh-push-tunnel')
+        ORDER BY S.id, SP.position DESC
+    """
+
+    h = _executeQuery(query)
+
+    prow = None
+    proxies = []
+
+    row = h.fetchone()
+    while True:
+        if prow is not None and (row is None or row[0] != prow[0]):
+            ret[prow[1]] = _getSSHMinion(
+                minion_id=prow[1],
+                proxies=proxies,
+                tunnel=prow[3],
+                ssh_push_port=int(prow[2]),
+            )
+            proxies = []
+        if row is None:
+            break
+        if row[4]:
+            proxies.append(row[4])
+        prow = row
+        row = h.fetchone()
+
+    cache.store("roster/uyuni", "minions", {"fp": cache_fp, "minions": ret})
+    log.trace("Uyuni DB roster:\n%s" % dump(ret))
```

Comment:
I hope I understood correct to make it the following way:
```
    if log.isEnabledFor(logging.TRACE):
        log.trace("Uyuni DB roster:\n%s", dump(ret))
```

But for some reason it doesn't work as supposed to be. It's easy to check with this modification:
```
    if log.isEnabledFor(logging.TRACE):
        log.error("Uyuni DB roster:\n%s", dump(ret))
```
In this case it will write the error in each case no matter which logging level is set.

Reply:
Yesterday I just meat to pass the args not yet formatted (`s/%/,/`), but today I realize this does not actually work here since `dump` will always be called before its result is passed to `log.trace`. Maybe the problem with `logging.TRACE` is that we never defined that log level. Salt has it defined, but it is not a default log level https://docs.python.org/3/howto/logging.html#logging-levels

Reply:
I've tried to change it to DEBUG, but the result is the same.

Reply:
Finally I left it as:
```
    if log.isEnabledFor(logging.TRACE):
        log.trace("Uyuni DB roster:\n%s", dump(ret))
```
But due to the bug in salt it goes inside conditional part in any logging level. Hope it will be fixed.